File: iOS/WatBackend/WatBackend/MapFactory.m

#import <Foundation/Foundation.h>
#import "MapFactory.h"
#import "Map.h"
#import "Location.h"
#import "Util.h"

@implementation MapFactory

// Pretend we're java Scanner.next()
void *scanNext(NSScanner *scanner, NSString **into){
    [scanner scanUpToCharactersFromSet:[NSCharacterSet whitespaceAndNewlineCharacterSet] intoString:into];
    return 0;
}

NSScanner *preprocess(NSString *str){
    NSMutableString *poststr = [[NSMutableString alloc] init];
    for(NSString *line in [str componentsSeparatedByString:@"\n"]){
        NSString *line2 = [line stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
        if([line2 length] == 0) continue;
        if([line2 characterAtIndex:0] == '#') continue;
        [poststr appendString:line2];
        [poststr appendString:@"\n"];
    }
    return [NSScanner scannerWithString:poststr];
}

void handleCommandLocation(Map *map, NSScanner *scanner){
    NSString *s;
    NSString *name; scanNext(scanner, &name);
    int pos_x; [scanner scanInt:&pos_x];
    int pos_y; [scanner scanInt:&pos_y];
    int num_floors = 1;
    int main_floor = 1;
    BOOL zero_indexed = false;
    BOOL selectable = true;
    while(!scanner.atEnd){
        scanNext(scanner, &s);
        if([s isEqual:@";"]) break;
        
        if([s isEqual:@"floors"]){
            [scanner scanInt:&num_floors];
        }
        if([s isEqual:@"main_floor"]){
            [scanner scanInt:&main_floor];
        }
        if([s isEqual:@"has_basement"]){
            zero_indexed = true;
        }
        if([s isEqual:@"unselectable"]){
            selectable = false;
        }
    }
    
    [map addBuilding:[[Building alloc] initWithName:name position:[[Waypoint alloc] initWithX:pos_x withY:pos_y] num_floors:num_floors main_floor:main_floor zero_indexed:zero_indexed selectable:selectable]];
}

void handleCommandPassiveLocation(Map *map, NSScanner *scanner){
    NSString *name; scanNext(scanner, &name);
    int pos_x; [scanner scanInt:&pos_x];
    int pos_y; [scanner scanInt:&pos_y];
    [map addPassiveLocation:[[Location alloc] initWithName:name withX:pos_x withY:pos_y active:false]];
}

void handleCommandPath(Map *map, NSScanner *scanner){
    NSString *name1; scanNext(scanner, &name1);
    NSString *name2; scanNext(scanner, &name2);
    
    Location *roughly_loc1 = [map getLocationByID:name1];
    Location *roughly_loc2 = [map getLocationByID:name2];
    
    NSMutableArray *connect_floors1 = [[NSMutableArray alloc] init];
    NSMutableArray *connect_floors2 = [[NSMutableArray alloc] init];
    
    int pathType = TYPE_OUTSIDE;
    
    NSMutableArray *waypoints = [[NSMutableArray alloc] init];
    [waypoints addObject:[roughly_loc1 position]];
     
    NSString *s;
    while(!scanner.atEnd){
        scanNext(scanner, &s);
        if([s isEqual:@";"]) break;
        
        if([s isEqual:@"p"]){
            int wx; [scanner scanInt:&wx];
            int wy; [scanner scanInt:&wy];
            [waypoints addObject:[[Waypoint alloc] initWithX:wx withY:wy]];
        }
        if([s isEqual:@"type"]){
            NSString *type_str; scanNext(scanner, &type_str);
            if([type_str isEqual:@"inside"]){
                pathType = TYPE_INSIDE;
            }
            if([type_str isEqual:@"indoor_tunnel"]){
                pathType = TYPE_INDOOR_TUNNEL;
            }
            if([type_str isEqual:@"underground_tunnel"]){
                pathType = TYPE_UNDERGROUND_TUNNEL;
            }
            if([type_str isEqual:@"briefly_outside"]){
                pathType = TYPE_BRIEFLY_OUTSIDE;
            }
        }
        if([s isEqual:@"connects"]){
            int c1; [scanner scanInt:&c1];
            int c2; [scanner scanInt:&c2];
            [connect_floors1 addObject:[NSNumber numberWithInt:c1]];
            [connect_floors2 addObject:[NSNumber numberWithInt:c2]];
        }
    }
    [waypoints addObject:[roughly_loc2 position]];
    
    if([connect_floors1 count] == 0){
        int main_floor1 = 1;
        int main_floor2 = 1;
        Building *build1 = [map getBuildingByID:name1];
        Building *build2 = [map getBuildingByID:name2];
        if(build1 != nil) main_floor1 = build1.main_floor;
        if(build2 != nil) main_floor2 = build2.main_floor;
        [connect_floors1 addObject:[NSNumber numberWithInt:main_floor1]];
        [connect_floors2 addObject:[NSNumber numberWithInt:main_floor2]];
    }
    
    for(int i=0; i<[connect_floors1 count]; i++){
        Location *loc1 = [map getLocationByID:makeBuildingAndFloor(name1, [[connect_floors1 objectAtIndex:i] intValue])];
        Location *loc2 = [map getLocationByID:makeBuildingAndFloor(name2, [[connect_floors2 objectAtIndex:i] intValue])];
        Path *path = [[Path alloc] initWithLocationA:loc1 withLocationB:loc2];
        [path setWaypointsInfo:waypoints];
        [path setPathType:pathType];
        [map addPath:path];
    }
}

Map *readMapFromScanner(NSScanner *scanner){
    Map *map = [[Map alloc] init];
    
    while(!scanner.atEnd){
        NSString *command;
        if([scanner scanString:@"location" intoString:&command]){
            handleCommandLocation(map, scanner);
        }else if([scanner scanString:@"passive_location" intoString:&command]){
            handleCommandPassiveLocation(map, scanner);
        }else if([scanner scanString:@"path" intoString:&command]){
            handleCommandPath(map, scanner);
        }else{
            scanner.scanLocation++;
        }
    }
    
    return map;
}

+ (Map*)readMapFromPath: (NSString*) path{
    
    NSError *error;
    NSString *content = [[NSString alloc] initWithContentsOfFile:path encoding:NSUTF8StringEncoding error:&error];
    if(content == nil){
        NSLog(@"%@\n", error);
    }
    
    NSScanner *stream = preprocess(content);
    Map *map = readMapFromScanner(stream);
    
    return map;
}

@end

File: android/WATisRain/RainBackend/src/main/java/com/lucky/watisrain/backend/MapFactory.java
package com.lucky.watisrain.backend;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

import com.lucky.watisrain.backend.data.Building;
import com.lucky.watisrain.backend.data.Location;
import com.lucky.watisrain.backend.data.Map;
import com.lucky.watisrain.backend.data.Path;
import com.lucky.watisrain.backend.data.Waypoint;

/**
 * MapFactory generates a Map object given external input data.
 */
public class MapFactory {
	
	
	/**
	 * Read a Map object given a Scanner pointed to the stream
	 */
	public static Map readMapFromScanner(Scanner scanner){
		
		// removes empty lines and comments
		scanner = preprocess(scanner);
		Map map = new Map();
		
		while(scanner.hasNext()){
			
			String command = scanner.next();
			
			// Delegate handling to sub-functions
			switch(command){
				case "location":
					handleCommandLocation(map, scanner);
					break;
				case "passive_location":
					handleCommandPassiveLocation(map, scanner);
					break;
				case "path":
					handleCommandPath(map, scanner);
					break;
				default:
					throw new RuntimeException("Invalid command: " + command);
			}
		}
		
		scanner.close();
		
		return map;
	}
	
	
	/*
	 * Location (actually a building)
	 */
	private static void handleCommandLocation(Map map, Scanner scanner){
		
		String name = scanner.next();
		int pos_x = scanner.nextInt();
		int pos_y = scanner.nextInt();
		int num_floors = 1;
		int main_floor = 1;
		boolean zero_indexed = false;
		boolean selectable = true;
		
		while(true){
			if(!scanner.hasNext()) break;
			String s = scanner.next();
			if(s.equals(";")) break;
			
			if(s.equals("floors")){
				num_floors = scanner.nextInt();
			}
			
			if(s.equals("main_floor")){
				main_floor = scanner.nextInt();
			}
			
			if(s.equals("has_basement")){
				zero_indexed = true;
			}
			
			if(s.equals("unselectable")){
				selectable = false;
			}
		}
		
		Building building = new Building(name, new Waypoint(pos_x, pos_y), num_floors, main_floor, zero_indexed, selectable);
		
		map.addBuilding(building);
	}
	
	
	/*
	 * Passive location. Simply [name] [x] [y].
	 */
	private static void handleCommandPassiveLocation(Map map, Scanner scanner){
		String name = scanner.next();
		int pos_x = scanner.nextInt();
		int pos_y = scanner.nextInt();
		map.addPassiveLocation(new Location(name,pos_x,pos_y, false));
	}
	
	
	/*
	 * Paths
	 */
	private static void handleCommandPath(Map map, Scanner scanner){
		
		String name1 = scanner.next();
		String name2 = scanner.next();
		
		// Figure out where we are, approximately first. This location will have the
		// correct waypoint, but may have the incorrect floor.
		Location roughly_loc1 = map.getLocationByID(name1);
		Location roughly_loc2 = map.getLocationByID(name2);
		
		// We can have 0 or more "connects" commands. Store them in a list.
		ArrayList<Integer> connect_floors1 = new ArrayList<>();
		ArrayList<Integer> connect_floors2 = new ArrayList<>();
		
		int pathType = Path.TYPE_OUTSIDE;
		
		// Read waypoints
		List<Waypoint> waypoints = new ArrayList<>();
		waypoints.add(roughly_loc1.getPostion());
		
		// Read until semicolon
		while(true){
			if(!scanner.hasNext()) break;
			String s = scanner.next();
			if(s.equals(";")) break;
			
			if(s.equals("p")){
				// add waypoint
				int wx = scanner.nextInt();
				int wy = scanner.nextInt();
				
				waypoints.add(new Waypoint(wx,wy));
			}
			
			if(s.equals("type")){
				String type_str = scanner.next();
				switch(type_str){
					case "inside":
						pathType = Path.TYPE_INSIDE;
						break;
					case "indoor_tunnel":
						pathType = Path.TYPE_INDOOR_TUNNEL;
						break;
					case "underground_tunnel":
						pathType = Path.TYPE_UNDERGROUND_TUNNEL;
						break;
					case "briefly_outside":
						pathType = Path.TYPE_BRIEFLY_OUTSIDE;
						break;
				}
			}
			
			if(s.equals("connects")){
				connect_floors1.add(scanner.nextInt());
				connect_floors2.add(scanner.nextInt());
			}
		}
		
		waypoints.add(roughly_loc2.getPostion());
		
		
		// No "connects" specified, then just link main floor to main floor
		if(connect_floors1.isEmpty()){
			int main_floor1 = 1;
			int main_floor2 = 1;
			Building build1 = map.getBuildingByID(name1);
			Building build2 = map.getBuildingByID(name2);
			if(build1 != null) main_floor1 = build1.getMainFloorNumber();
			if(build2 != null) main_floor2 = build2.getMainFloorNumber();
			connect_floors1.add(main_floor1);
			connect_floors2.add(main_floor2);
		}
		
		
		
		// Add a path for each "connects"
		for(int i=0; i<connect_floors1.size(); i++){
			Location loc1 = map.getLocationByID(Util.makeBuildingAndFloor(name1, connect_floors1.get(i)));
			Location loc2 = map.getLocationByID(Util.makeBuildingAndFloor(name2, connect_floors2.get(i)));

			Path path = new Path(loc1,loc2);
			path.setWaypoints(waypoints);
			path.setPathType(pathType);
			
			map.addPath(path);
		}
	}
	
	
	
	// Preprocess to remove blank lines and comment lines
	private static Scanner preprocess(Scanner prescanner){
		
		StringBuilder sb = new StringBuilder();
		while(prescanner.hasNextLine()){
			String line = prescanner.nextLine();
			if(line.trim().isEmpty() || line.charAt(0) == '#') continue;
			sb.append(line + "\n");
		}
		
		// Actual scanner, don't worry about extraneous stuff anymore
		return new Scanner(sb.toString());
	}
	
	
	

	/**
	 * Create a Map object from a specified File
	 */
	public static Map readMapFromFile(File infile)
			throws FileNotFoundException{
		
		return readMapFromScanner(
				new Scanner(
					new FileInputStream(infile)));
		
	}
	
	
	public static Map readMapFromStream(InputStream stream){
		return readMapFromScanner(new Scanner(stream));
	}
	
	
}
